src/tensors/levels/sparsecoolevels.jl

```diff
@@ -33,34 +37,73 @@ SparseCOO (0.0) [1:3,1:3]
 
-SparseCOOLevel{N, Ti}(lvl, args...) where {N, Ti} = SparseCOOLevel{N, Ti, Int}(lvl, args...)
-SparseCOOLevel{N, Ti, Tp}(lvl::Lvl, args...) where {N, Ti, Tp, Lvl} =
-    SparseCOOLevel{N, Ti, Tp, Tuple{(Vector{ti} for ti in Ti.parameters)...}, Lvl}(lvl, args...)
+SparseCOOLevel{N, Ti}(lvl, args...) where {N, Ti} =
+    SparseCOOLevel{N,
+        Ti,
+        postype(typeof(lvl)),
+        Tuple{((memory_type(typeof(lvl))){ti, 1} for ti in Ti.parameters)...},
+        (memory_type(typeof(lvl))){postype(typeof(lvl)), 1},
+        typeof(lvl)}(lvl, args...)
+SparseCOOLevel{N, Ti}(lvl) where {N, Ti} =
+    SparseCOOLevel{N,
+        Ti,
+        postype(typeof(lvl)),
+        Tuple{((memory_type(typeof(lvl))){ti, 1} for ti in Ti.parameters)...},
+        (memory_type(typeof(lvl))){postype(typeof(lvl)), 1},
+        typeof(lvl)}(lvl)
+
+SparseCOOLevel{N, Ti, Tp}(lvl, args...) where {N, Ti, Tp} =
+    SparseCOOLevel{N, Ti, Tp, Tuple{(memory_type(typeof(lvl)){ti, 1} for ti in Ti.parameters)...}, memory_type(typeof(lvl)){Tp, 1}, typeof(lvl)}(lvl, args...)
+
+SparseCOOLevel{N, Ti, Tp, Tbl, VTp, Lvl}(lvl) where {N, Ti, Tp, Tbl, VTp, Lvl} =
+    SparseCOOLevel{N, Ti, Tp, Tbl, VTp, Lvl}(lvl, ((zero(ti) for ti in Ti.parameters)...,))
+SparseCOOLevel{N, Ti, Tp, Tbl, VTp, Lvl}(lvl, shape) where {N, Ti, Tp, Tbl, VTp, Lvl} =
+    SparseCOOLevel{N, Ti, Tp, Tbl, VTp, Lvl}(lvl, Ti(shape), ((ti() for ti in Tbl.parameters)...,), VTp(Tp[1]))
+
 
-SparseCOOLevel{N, Ti, Tp, Tbl, Lvl}(lvl) where {N, Ti, Tp, Tbl, Lvl} =
-    SparseCOOLevel{N, Ti, Tp, Tbl, Lvl}(lvl, ((zero(ti) for ti in Ti.parameters)...,))
-SparseCOOLevel{N, Ti, Tp, Tbl, Lvl}(lvl, shape) where {N, Ti, Tp, Tbl, Lvl} =
-    SparseCOOLevel{N, Ti, Tp, Tbl, Lvl}(lvl, Ti(shape), ((Vector{ti}() for ti in Ti.parameters)...,), Tp[1])
+# SparseCOOLevel{N, Ti}(lvl) where {N, Ti} = SparseCOOLevel{N, Ti, postype(typeof(lvl)), Tuple{(memory_type(typeof(lvl)){ti, 1} for ti in Ti.parameters)...}, memory_type(typeof(lvl)){postype(typeof(lvl)), 1}, typeof(lvl)}(lvl)
 
 Base.summary(lvl::SparseCOOLevel{N}) where {N} = "SparseCOO{$N}($(summary(lvl.lvl)))"
 similar_level(lvl::SparseCOOLevel{N}) where {N} = SparseCOOLevel{N}(similar_level(lvl.lvl))
 similar_level(lvl::SparseCOOLevel{N}, tail...) where {N} = SparseCOOLevel{N}(similar_level(lvl.lvl, tail[1:end-N]...), (tail[end-N+1:end]...,))
 
-pattern!(lvl::SparseCOOLevel{N, Ti, Tp}) where {N, Ti, Tp} = 
-    SparseCOOLevel{N, Ti, Tp}(pattern!(lvl.lvl), lvl.shape, lvl.tbl, lvl.ptr)
+
+
+function memory_type(::Type{SparseCOOLevel{N, Ti, Tp, Tbl, VTp, Lvl}}) where {N, Ti, Tp, Tbl, VTp, Lvl}
+    return containertype(VTp)
+end
+
+function postype(::Type{SparseCOOLevel{N, Ti, Tp, Tbl, VTp, Lvl}}) where {N, Ti, Tp, Tbl, VTp, Lvl}
+    return Tp
+end
+
+function indextype(::Type{SparseCOOLevel{N, Ti, Tp, Tbl, VTp, Lvl}}) where {N, Ti, Tp, Tbl, VTp, Lvl}
+    return indextype(Ti)
```

Comment:
A question about indextype. Should this perhaps just return the top index?

Reply:
rather than tuple{Int}?

Reply:
`indextype` type on a tuple of types does this.

Reply:
wait, so is the argument to indextype a level or is it an index?

Reply:
should we just call `first(Ti.parameters)`?

Reply:
or `last(Ti.parameters)`?

Reply:
> wait, so is the argument to indextype a level or is it an index?

Either. Indextype on a level typically ends up calling index type on a tuple of types.

Reply:
I felt this was a clean way to deal with the variety of things that could be put there.